asllib/doc/TopLevel.tex

```diff
@@ -0,0 +1,104 @@
+        \{\ \LexicalError, \BuildErrorConfig\ \}    & \cup\\
+        \overname{\TTypeError}{\TypeErrorConfig}    & \cup\\
+        \overname{\TDynError}{\DynErrorConfig}      &
+    \end{array}
+\right)
+\]
+accepts a textual description of a specification in $\vspectext$ and a textual description of the standard library in $\vstdtext$.
+The descriptions are statically checked for validity.
+If found invalid, an error configuration corresponding to the phase where the error
+exists is returned --- scanning, parsing, or AST building.
+I found valid, an AST is built and semantically evaluated, yielding an integer return code $\vret$ and an execution graph $\vg$,
+or a dynamic error.
+
+\subsubsection{TopLevelRule.CheckAndInterpret}
+\subsubsection{Prose}
+All of the following apply:
+\begin{itemize}
+    \item \Proseaslscan{$\vstdtext$}{$\spectoken$}{$\vstdtokens$\ProseTerminateAs{\LexicalError}};
+    \item \Proseaslparse{$\vstdtokens$}{$\vstdparse$\ProseTerminateAs{\ParseError}};
+    \item \Prosebuildast{$\vstdparse$}{$\vstdast$\ProseTerminateAs{\BuildErrorConfig}};
+    \item define $\vstdasbuiltin$ by applying $\setbuiltin$ to each top-level declaration in\\
+          $\vstdast$\ProseOrTypeError;
+    %%%%%
+    \item \Proseaslscan{$\vspectext$}{$\spectoken$}{$\vspectokens$\ProseTerminateAs{\LexicalError}};
+    \item \Proseaslparse{$\vspectokens$}{$\vspecparse$\ProseTerminateAs{\ParseError}};
+    \item \Prosebuildast{$\vspecparse$}{$\vspecast$\ProseTerminateAs{\BuildErrorConfig}};
+    %%%%%
+    \item define $\vuntypedast$ as the concatenation of the AST $\vstdasbuiltin$ and
+          the AST $\vspecast$ (both are lists of $\decl$);
+    \item \Prosetypecheckast{$\vuntypedast$}{empty static global environment}{$\vtypedast$}{$\tenv$\ProseOrTypeError};
+    \item \Proseevalspec{$\vtypedast$}{$\tenv$}{the integer $\vret$ and the execution graph $\vg$\ProseOrError}.
+\end{itemize}
+
+\subsubsection{Formally}
+\begin{mathpar}
+\inferrule{
+    \aslscan(\spectoken, \vstdtext) \scanarrow \vstdtokens \terminateas \LexicalError\\\\
+    \aslparse(\vstdtokens) \parsearrow \vstdparse \terminateas \ParseError\\\\
+    \buildast(\vstdparse) \astarrow \vstdast \terminateas \BuildErrorConfig\\\\
+    \vi \in \listrange(\vstdast): \setbuiltin(\vstdast_i) \typearrow \vstddeclast_i \OrTypeError\\\\
+    \vstdasbuiltin \eqdef [\vi \in \listrange(\vstdast): \vstddeclast_i]\\
+    \aslscan(\spectoken, \vspectext) \scanarrow \vspectokens \terminateas \LexicalError\\\\
+    \aslparse(\vspectokens) \parsearrow \vspecparse \terminateas \ParseError\\\\
+    \buildast(\vspecparse) \astarrow \vstdast \terminateas \BuildErrorConfig\\\\
+    \vuntypedast \eqdef \vstdasbuiltin \concat \vspecast\\
+    \typecheckast(G^{\emptytenv}, \vuntypedast) \typearrow (\vtypedast, \tenv) \OrTypeError\\\\
+    \evalspec(\tenv, \vtypedast) \evalarrow (\nvint(\vret), \vg) \OrDynError
+}{
+    \checkandinterpret(\vspectext, \vstdtext) \longrightarrow (\nvint(\vret), \vg)
```

Comment:
It feels like there should be no `\nvint` here in the conclusion

Reply:
Why? We want the `main` function to return an integer return value.

Reply:
Ah sorry, I misread the type declaration for `check_and_interpret` as returning a `\mathbb{Z}` not a `\mathcal{Z}` - my mistake. I suppose there's an argument it could be `\mathbb{Z}`?

Reply:
I thought about returning a mathematical integer vs. a semantic value. I opted for the latter, since I think it may later be useful when we try to connect parallel specifications via the memory model.